x-pack/plugins/lens/public/drag_drop/drag_drop.tsx

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useContext } from 'react';
+import { DragContext } from './providers';
+
+type DroppableEvent = React.DragEvent<HTMLElement>;
+
+/**
+ * A function that handles a drop event.
+ */
+type DropHandler = (item: unknown) => void;
+
+/**
+ * The argument to the DragDrop component.
+ */
+interface Props {
+  /**
+   * The CSS class(es) for the root element.
+   */
+  className?: string;
+
+  /**
+   * The event handler that fires when an item
+   * is dropped onto this DragDrop component.
+   */
+  onDrop?: DropHandler;
+
+  /**
+   * The value associated with this item, if it is draggable.
+   * If this component is dragged, this will be the value of
+   * "dragging" in the root drag/drop context.
+   */
+  value?: unknown;
+
+  /**
+   * The React children.
+   */
+  children: React.ReactNode;
+
+  /**
+   * Indicates whether or not the currently dragged item
+   * can be dropped onto this component.
+   */
+  droppable?: boolean;
+
+  /**
+   * Indicates whether or not this component is draggable.
+   */
+  draggable?: boolean;
+}
+
+/**
+ * A draggable / droppable item. Items can be both draggable and droppable at
+ * the same time.
+ *
+ * @param props
+ */
+export function DragDrop(props: Props) {
+  const { dragging, setDragging } = useContext(DragContext);
+  const [state, setState] = useState({ isActive: false });
+  const { className, onDrop, value, children, droppable, draggable } = props;
+  const isDragging = draggable && value === dragging;
+
+  const classSuffix =
+    ' lnsDragDrop' +
+    (droppable ? ' lnsDragDrop-isDropTarget' : '') +
```

Comment:
You should think about using the `classNames` method here so this whole block can be re-written as:

``` jsx
const classes = classNames(
  'lnsDragDrop',
  {
    'lnsDragDrop-isDropTarget': droppable,
    'lnsDragDrop-isActiveDropTarget': droppable && state.isActive,
    'lnsDragDrop-isDragging': isDragging
  },
  className
);
```

which is a lot easier to follow and won't leave odd spacing between class names.

Reply:
@chrisdavies related to my classnames comment down below - as cchaos is aso suggesting it I think it’s worth it :-)